sqlglot/optimizer/annotate_expression_types.py

```diff
@@ -0,0 +1,207 @@
+
+    annotator = annotators.get(type(expression))
+    if not annotator:
+        raise OptimizeError(f"Unable to annotate expression of type: {type(expression)}")
+
+    return annotator(expression, schema, annotators)
+
+
+def _coerce(type1, type2):
+    return type2 if type2 in PRECEDED_BY[type1] else type1
+
+
+def _annotate_cast(expr, schema, annotators):
+    expr.type = expr.args["to"].this
+
+    _annotate(expr.this, schema, annotators)
+    _annotate(expr.args["to"], schema, annotators)
+
+    return expr
+
+
+def _annotate_cache(expr, schema, annotators):
+    _annotate(expr.expression, schema, annotators)
+    _annotate(expr.args.get("with"), schema, annotators)
+
+    return expr
+
+
+def _annotate_data_type(expr, schema, annotators):
+    expr.type = expr.this
+
+    for expression in expr.expressions:
+        _annotate(expression, schema, annotators)
+
+    return expr
+
+
+def _annotate_select(expression, schema, annotators):
+    for value in expression.args.values():
+        for v in ensure_list(value):
+            _annotate(v, schema, annotators)
+
+    return expression
+
+
+def _annotate_paren(expression, schema, annotators):
+    _annotate(expression.this, schema, annotators)
+
+    expression.type = expression.this.type
+
+    return expression
+
+
+def _annotate_binary(expression, schema, annotators):
+    _annotate(expression.left, schema, annotators)
+    _annotate(expression.right, schema, annotators)
+
+    expression.type = _coerce(expression.left.type, expression.right.type)
+
+    return expression
+
+
+def _annotate_connector(expr, schema, annotators):
+    _annotate(expression.left, schema, annotators)
+    _annotate(expression.right, schema, annotators)
+
+    expression.type = exp.DataType.Type.BOOLEAN
+
+    return expression
+
+
+def _annotate_literal(expression, schema, annotators):
+    if expression.is_string:
+        expression.type = exp.DataType.Type.VARCHAR
+    elif expression.is_int:
+        expression.type = exp.DataType.Type.INT
+    else:
+        expression.type = exp.DataType.Type.DOUBLE
+
+    return expression
+
+
+def _annotate_boolean(expression, schema, annotators):
+    expression.type = exp.DataType.Type.BOOLEAN
+    return expression
+
+
+def _set_precedence(types):
+    """
+    Sets the correct precedence for a set of related types.
+
+    Example:
+        >>> PRECEDED_BY = {}
+        >>> types = [exp.DataType.Type.DOUBLE, exp.DataType.Type.FLOAT]
+        >>> _set_precedence(types)
+        >>> print(PRECEDED_BY)
+        {exp.DataType.Type.DOUBLE: set(), exp.DataType.Type.FLOAT: exp.DataType.Type.DOUBLE}
+
+    Args:
+        types (list[exp.DataType.Type]): types, in descending order w.r.t their relative precedence
+    """
+
+    PRECEDED_BY[types[0]] = set()
+    for t1, t2 in zip(types[:-1], types[1:]):
+        PRECEDED_BY[t2] = {t1}.union(PRECEDED_BY[t1])
+
+
+def _create_precedence_mapping():
```

Comment:
you should not be modifying constants like this.  in functions.

PRECEDED_BY = ... in the global scope,

and then if someone passes in a new one, create a new instance